app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/Source.php

```diff
@@ -234,6 +251,9 @@ protected function _prepareMultiselectIndex($entityIds = null, $attributeId = nu
             $options[$row['attribute_id']][$row['option_id']] = true;
         }
 
+        // Include custom source model options
+        $options = $this->getMultiSelectAttributeWithSourceModels($attrIds, $options);
```

Comment:
@sivaschenko In order to do this the only option I can think of is to use [array union](http://php.net/manual/en/language.operators.array.php) to avoid a reindex of the `$options` array and maintain the keys as they are numeric. Do you have any other recommendations to approach this? If not then I'll push an update using union.

Reply:
What do you think about `$options = array_replace_recursive($options, $this->getMultiSelectAttributeWithSourceModels($attrIds));`

Reply:
Ah, I was thinking that would result in the same issue as `array_merge` but it looks like it won't. I'll move forward with that. Thanks!